File: ex-14/ex-14_python/renderizador.py
from vetor2f import Vetor2f
from cor import Cor
from vertex_array import VertexArray, LineStrip
from circle_shape import CircleShape


class Renderizador:
    def __init__(self, janela, larguraEspaco, alturaEspaco):
        self.janela = janela
        self.escalaX = 0.0
        self.escalaY = 0.0
        self.atualizarEscala(larguraEspaco, alturaEspaco)

    def atualizarEscala(self, larguraEspaco, alturaEspaco):
        tamanhoJanela = self.janela.getSize()
        self.escalaX = tamanhoJanela[0] / larguraEspaco
        self.escalaY = tamanhoJanela[1] / alturaEspaco

    def converterParaTela(self, posicaoEspaco):
        x = float(posicaoEspaco.getX() * self.escalaX)
        y = float(posicaoEspaco.getY() * self.escalaY)
        return Vetor2f(x, y)

    def renderizar(self, simulacao):
        self.janela.clear(Cor.Black())

        corpos = simulacao.getCorpos()

        for corpo in corpos:
            trajetoria = corpo.getTrajetoria()

            if len(trajetoria) > 1:
                linhas = VertexArray(LineStrip, len(trajetoria))
                for i in range(len(trajetoria)):
                    linhas[i].position = trajetoria[i]
                    corTrajetoria = corpo.getCor()
                    corTrajetoria.a = 100
                    linhas[i].color = corTrajetoria
                self.janela.draw(linhas)

            posicaoTela = self.converterParaTela(corpo.getPosicao())

            circulo = CircleShape(float(corpo.getRaio()))
            circulo.setFillColor(corpo.getCor())
            circulo.setOrigin(Vetor2f(float(corpo.getRaio()), float(corpo.getRaio())))
            circulo.setPosition(posicaoTela)

            self.janela.draw(circulo)

        self.janela.display()

File: ex-14/ex-14_java/Renderizador.java
import java.util.List;

public class Renderizador {
    private JanelaRender janela;
    private double escalaX;
    private double escalaY;

    public Renderizador(JanelaRender janela, double larguraEspaco, double alturaEspaco) {
        this.janela = janela;
        atualizarEscala(larguraEspaco, alturaEspaco);
    }

    public void atualizarEscala(double larguraEspaco, double alturaEspaco) {
        int[] tamanhoJanela = janela.getSize();
        escalaX = tamanhoJanela[0] / larguraEspaco;
        escalaY = tamanhoJanela[1] / alturaEspaco;
    }

    private Vetor2f converterParaTela(Vetor2D posicaoEspaco) {
        float x = (float) (posicaoEspaco.getX() * escalaX);
        float y = (float) (posicaoEspaco.getY() * escalaY);
        return new Vetor2f(x, y);
    }

    public void renderizar(Simulacao simulacao) {
        janela.clear(Cor.Black());

        List<CorpoCeleste> corpos = simulacao.getCorpos();

        for (CorpoCeleste corpo : corpos) {
            List<Vetor2f> trajetoria = corpo.getTrajetoria();

            if (trajetoria.size() > 1) {
                VertexArray linhas = new VertexArray(VertexArray.LineStrip, trajetoria.size());
                for (int i = 0; i < trajetoria.size(); i++) {
                    linhas.get(i).position = trajetoria.get(i);
                    Cor corTrajetoria = corpo.getCor();
                    corTrajetoria.a = 100;
                    linhas.get(i).color = corTrajetoria;
                }
                janela.draw(linhas);
            }

            Vetor2f posicaoTela = converterParaTela(corpo.getPosicao());

            CircleShape circulo = new CircleShape((float) corpo.getRaio());
            circulo.setFillColor(corpo.getCor());
            circulo.setOrigin(new Vetor2f((float) corpo.getRaio(), (float) corpo.getRaio()));
            circulo.setPosition(posicaoTela);

            janela.draw(circulo);
        }

        janela.display();
    }
}
